File: leetcode-submissions/42_Trapping_Rain_Water.ts
function trap(height: number[]): number {
    const n :number = height.length;
    const maxLeftHeights :number[]=new Array(n);
    const maxRightHeights :number[]=new Array(n);
    let maxLeftHeight = Number.MIN_SAFE_INTEGER;
    let maxRightHeight= Number.MIN_SAFE_INTEGER;
    for(let i:number=0;i<n;i++){
        maxLeftHeight= Math.max(maxLeftHeight, height[i]);
        maxLeftHeights[i]=maxLeftHeight;

    }
    for(let i:number = n-1;i>=0;i--){
        maxRightHeight= Math.max(maxRightHeight, height[i]);
        maxRightHeights[i]= maxRightHeight;
    }
    let ans :number =0;

    for(let i:number =0;i<n;i++){
        ans+=(Math.min(maxRightHeights[i],maxLeftHeights[i])-height[i]);
    }
    return ans ;
};

File: leetcode-submissions/42_Trapping_Rain_Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        // i have an appraoch that can able to solve that question now solving this question by uisng the two arrays right maximum and left maximum
        vector<int>leftmax(height.size());
        vector<int>rightmax(height.size());
        int leftmaxele=INT_MIN;
        for(int i=0;i<height.size();i++){
            leftmaxele=max(leftmaxele,height[i]);
            leftmax[i]=max(leftmaxele,height[i]);

        } 
        int rightmaxele=INT_MIN;
        for(int i=height.size()-1;i>=0;i--){
            rightmaxele=max(rightmaxele,height[i]);
            rightmax[i]=max(rightmaxele,height[i]);

        }
        int ans=0;
        for(int i=0;i<height.size();i++){
            ans+=(min(rightmax[i],leftmax[i])-height[i]);
        }
        return ans;
    }
};